### PR TITLE
i#3759: Fix mmap handling conflicts

### LINKED ISSUE #3759: Massive test failures on 64-bit and 32-bit when running in debug.
On a new (G)linux system 

cat /proc/version 
Linux version 4.19.37-5rodete1-amd64 (glinux-team@google.com) (gcc version 8.0.1 20180414 (experimental) [trunk revision 259383] (Debian 8-20180414-1)) #1 SMP Debian 4.19.37-5rodete1 (2019-06-20 > 2018)

I am now getting a whole barrage of test failures, in particular when running debug:

```
The following tests FAILED:
         67 - code_api|linux.syscall_pwait (Failed)
         80 - code_api|linux.sigsuspend (Failed)
         81 - code_api|linux.signest (Failed)
        119 - code_api|client.modules (Failed)
        130 - code_api|client.annotation-concurrency (Failed)
        131 - code_api|client.annotation-concurrency.full-decode (Failed)
        132 - code_api|client.annotation-concurrency.full-decode.tiny-bb (Failed)
        133 - code_api|client.annotation-concurrency.bb-truncate-1 (Failed)
        134 - code_api|client.annotation-concurrency.bb-truncate-2 (Failed)
        135 - code_api|client.annotation-concurrency.prof-pcs (Failed)
        136 - code_api|client.annotation-concurrency-opt (Failed)
        137 - code_api|client.annotation-concurrency-opt.full-decode (Failed)
        138 - code_api|client.annotation-detection (Failed)
        139 - code_api|client.annotation-detection.full-decode (Failed)
        140 - code_api|client.annotation-detection.full-decode.tiny-bb (Failed)
        141 - code_api|client.annotation-detection.bb-truncate-1 (Failed)
        142 - code_api|client.annotation-detection.bb-truncate-2 (Failed)
        143 - code_api|client.annotation-detection-opt (Failed)
        157 - code_api|client.events (Failed)
        158 - code_api|client.events_cpp (Failed)
        175 - code_api|client.null_instrument (Failed)
        181 - code_api|client.drmgr-test (Failed)
        188 - code_api|client.drwrap-test (Failed)
        189 - code_api|client.drwrap-test-callconv (Failed)
        190 - code_api|client.drsyms-test (Failed)
        198 - code_api|client.destructor (Failed)
        231 - code_api|tool.drcov.fib (Failed)
        244 - code_api|tool.drcachesim.threads (Failed)
        245 - code_api|tool.drcachesim.threads-with-config-file (Failed)
        246 - code_api|tool.drcachesim.TLB-threads (Failed)
        248 - code_api|tool.drcachesim.miss_analyzer (Failed)
        308 - code_api|api.static_sideline_FLAKY (Failed)
        311 - code_api|api.static_maps_mixup_novars_FLAKY (Failed)
        314 - code_api,opt_memory|client.events (Failed)
        315 - code_api,opt_speed|client.events (Failed)
        316 - code_api,thread_private|client.events (Failed)
        317 - code_api,disable_traces|client.events (Failed)
        318 - code_api,thread_private,disable_traces|client.events (Failed)
        319 - code_api,no_early_inject|client.events (Failed)
Errors while running CTest
```

As a side note, the number of release test failures also has increased:

The following tests FAILED:
        179 - code_api|client.drmgr-test (Failed)
        229 - code_api|tool.drcov.fib (Failed)
        303 - code_api|api.static_signal (Failed)
        305 - code_api|api.static_sideline_FLAKY (Failed)
        308 - code_api|api.static_maps_mixup_novars_FLAKY (Failed)
Errors while running CTest

The change appears to be something about the system, not DynamoRIO itself. Rolling back to known better states leads to the same results.

### PR DESCRIPTION
**Pull request by derekbruening**
Fixes problems with pre-syscall versus post-syscall mmap handling
which can lead to asserts.  Adds a check-only version of
app_memory_protection_change() for use in pre-syscall, delaying all
vmarea updates to post-syscall.  Adds protection change checking
inside app_memory_allocation(), simplifying caller usage and providing
broader coverage of mmap behavior.

Adds a vmm attribute to all_memory_areas and disallows merging across
that boundary, to match kernel behavior and avoid mismatch warnings
with the above fix in place which does extra memory querying on each
mmap.

Adds some mmap sequences to linux.mmap to provide some test coverage
for these particular page protection change orderings.

Fixes #3759

### CODE REVIEW

File: core/lib/instrument.c [hunk 1]
```diff
@@ -2974,7 +2974,8 @@ raw_mem_alloc(size_t size, uint prot, void *addr, dr_alloc_flags_t flags)
             : (TEST(DR_ALLOC_COMMIT_ONLY, flags) ? RAW_ALLOC_COMMIT_ONLY : 0);
 #    endif
         if (IF_WINDOWS(TEST(DR_ALLOC_COMMIT_ONLY, flags) &&) addr != NULL &&
-            !app_memory_pre_alloc(get_thread_private_dcontext(), addr, size, prot, false))
+            !app_memory_pre_alloc(get_thread_private_dcontext(), addr, size, prot, false,
+                                  true /*update*/, false /*!image*/))
```
Comment (hgreving2304) [inline core/lib/instrument.c]: Maybe {}, even though no multi-line body?

Comment (derekbruening) [reply core/lib/instrument.c]: Not sure what the reasoning is, but I can add it.  Done.

**Comment by derekbruening** [comment]
Strange: the Travis build completed yet the wrapper page does not show the completion and here it claims it's still in progress.

**Comment by derekbruening** [comment]
TestMemProtChg_FLAKY seems to be failing when it did not before.  Investigating.

**Pull request merged**